Invocation_NetworkSplitTest

https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/11642/testReport/junit/com.hazelcast.spi.impl.operationservice.impl/Invocation_NetworkSplitTest/testWaitingInvocations_whenNodePartiallySplitFromCluster/

com.hazelcast.spi.impl.operationservice.impl.Invocation_NetworkSplitTest.testWaitingInvocations_whenNodePartiallySplitFromCluster

```
Error Message

java.util.concurrent.ExecutionException: com.hazelcast.spi.exception.CallerNotMemberException: Not Member! this: [127.0.0.1]:5002, caller: [127.0.0.1]:5003, partitionId: 0, operation: com.hazelcast.spi.impl.operationservice.impl.Invocation_NetworkSplitTest$AlwaysBlockingOperation, service: AlwaysBlockingOperationService
Stacktrace

java.lang.AssertionError: java.util.concurrent.ExecutionException: com.hazelcast.spi.exception.CallerNotMemberException: Not Member! this: [127.0.0.1]:5002, caller: [127.0.0.1]:5003, partitionId: 0, operation: com.hazelcast.spi.impl.operationservice.impl.Invocation_NetworkSplitTest$AlwaysBlockingOperation, service: AlwaysBlockingOperationService
	at org.junit.Assert.fail(Assert.java:88)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation_NetworkSplitTest.testWaitingInvocations_whenNodeSplitFromCluster(Invocation_NetworkSplitTest.java:109)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation_NetworkSplitTest.testWaitingInvocations_whenNodePartiallySplitFromCluster(Invocation_NetworkSplitTest.java:67)
```

Reopening. Failure in PR builder:
```
com.hazelcast.spi.impl.operationservice.impl.Invocation_NetworkSplitTest.testWaitNotifyService_whenNodeSplitFromCluster

Stacktrace
java.lang.AssertionError: expected:<4> but was:<3>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation_NetworkSplitTest.testWaitNotifyService_whenNodeSplitFromCluster(Invocation_NetworkSplitTest.java:167)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation_NetworkSplitTest.testWaitNotifyService_whenNodeSplitFromCluster(Invocation_NetworkSplitTest.java:114)
```

https://hazelcast-l337.ci.cloudbees.com/job/new-lab-fast-pr/14172/testReport/com.hazelcast.spi.impl.operationservice.impl/Invocation_NetworkSplitTest/testWaitNotifyService_whenNodeSplitFromCluster/

--- a/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/Invocation_NetworkSplitTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/spi/impl/operationservice/impl/Invocation_NetworkSplitTest.java
@@ -20,15 +20,12 @@ import com.hazelcast.config.Config;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.core.MemberLeftException;
 import com.hazelcast.instance.Node;
-import com.hazelcast.instance.TestUtil;
 import com.hazelcast.internal.cluster.impl.ClusterServiceImpl;
 import com.hazelcast.spi.AbstractWaitNotifyKey;
 import com.hazelcast.spi.BlockingOperation;
 import com.hazelcast.spi.ExceptionAction;
 import com.hazelcast.spi.Operation;
-import com.hazelcast.spi.OperationService;
 import com.hazelcast.spi.WaitNotifyKey;
-import com.hazelcast.spi.impl.NodeEngineImpl;
 import com.hazelcast.spi.impl.operationparker.impl.OperationParkerImpl;
 import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.test.AssertTask;
@@ -37,12 +34,10 @@ import com.hazelcast.test.HazelcastTestSupport;
 import com.hazelcast.test.TestHazelcastInstanceFactory;
 import com.hazelcast.test.annotation.ParallelTest;
 import com.hazelcast.test.annotation.QuickTest;
-import org.junit.Assert;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import java.util.concurrent.Callable;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
@@ -66,35 +61,33 @@ public class Invocation_NetworkSplitTest extends HazelcastTestSupport {
         testWaitingInvocations_whenNodeSplitFromCluster(action);
     }
 
-    private void testWaitingInvocations_whenNodeSplitFromCluster(SplitAction splitAction) throws Exception {
+    private void testWaitingInvocations_whenNodeSplitFromCluster(SplitAction action) {
         Config config = createConfig();
         TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(3);
-        HazelcastInstance hz1 = factory.newHazelcastInstance(config);
-        HazelcastInstance hz2 = factory.newHazelcastInstance(config);
-        HazelcastInstance hz3 = factory.newHazelcastInstance(config);
-
-        Node node1 = TestUtil.getNode(hz1);
-        Node node2 = TestUtil.getNode(hz2);
-        Node node3 = TestUtil.getNode(hz3);
+        final HazelcastInstance hz1 = factory.newHazelcastInstance(config);
+        final HazelcastInstance hz2 = factory.newHazelcastInstance(config);
+        final HazelcastInstance hz3 = factory.newHazelcastInstance(config);
 
         warmUpPartitions(hz1, hz2, hz3);
         int partitionId = getPartitionId(hz2);
 
-        NodeEngineImpl nodeEngine3 = node3.getNodeEngine();
-        OperationService operationService3 = nodeEngine3.getOperationService();
         Operation op = new AlwaysBlockingOperation();
-        Future<Object> future = operationService3.invokeOnPartition("", op, partitionId);
+        Future<Object> future = getNodeEngineImpl(hz3).getOperationService().invokeOnPartition("", op, partitionId);
 
-        // just wait a little to make sure
-        // operation is landed on wait-queue
-        sleepSeconds(1);
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run() {
+                final OperationParkerImpl waitNotifyService3 = (OperationParkerImpl) getNodeEngineImpl(hz2).getOperationParker();
+                assertEquals(1, waitNotifyService3.getTotalParkedOperationCount());
+            }
+        }, 20);
 
         // execute the given split action
-        splitAction.run(node1, node2, node3);
+        action.run(getNode(hz1), getNode(hz2), getNode(hz3));
 
         // Let node3 detect the split and merge it back to other two.
-        ClusterServiceImpl clusterService3 = node3.getClusterService();
-        clusterService3.merge(node1.address);
+        ClusterServiceImpl clusterService3 = (ClusterServiceImpl) getClusterService(hz3);
+        clusterService3.merge(getAddress(hz1));
 
         assertClusterSizeEventually(3, hz1, hz2, hz3);
 
@@ -120,33 +113,27 @@ public class Invocation_NetworkSplitTest extends HazelcastTestSupport {
         testWaitNotifyService_whenNodeSplitFromCluster(action);
     }
 
-    private void testWaitNotifyService_whenNodeSplitFromCluster(SplitAction action) throws Exception {
+    private void testWaitNotifyService_whenNodeSplitFromCluster(SplitAction action) {
         Config config = createConfig();
-        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(5);
-        HazelcastInstance hz1 = factory.newHazelcastInstance(config);
-        HazelcastInstance hz2 = factory.newHazelcastInstance(config);
-        HazelcastInstance hz3 = factory.newHazelcastInstance(config);
-
-        final Node node1 = TestUtil.getNode(hz1);
-        Node node2 = TestUtil.getNode(hz2);
-        Node node3 = TestUtil.getNode(hz3);
+        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(4);
+        final HazelcastInstance hz1 = factory.newHazelcastInstance(config);
+        final HazelcastInstance hz2 = factory.newHazelcastInstance(config);
+        final HazelcastInstance hz3 = factory.newHazelcastInstance(config);
 
         warmUpPartitions(hz1, hz2, hz3);
         int partitionId = getPartitionId(hz3);
 
-        NodeEngineImpl nodeEngine1 = node1.getNodeEngine();
-        OperationService operationService1 = nodeEngine1.getOperationService();
-        operationService1.invokeOnPartition("", new AlwaysBlockingOperation(), partitionId);
+        getNodeEngineImpl(hz1).getOperationService().invokeOnPartition("", new AlwaysBlockingOperation(), partitionId);
 
-        final OperationParkerImpl waitNotifyService3 = (OperationParkerImpl) node3.getNodeEngine().getOperationParker();
-        assertEqualsEventually(new Callable<Integer>() {
+        final OperationParkerImpl waitNotifyService3 = (OperationParkerImpl) getNodeEngineImpl(hz3).getOperationParker();
+        assertTrueEventually(new AssertTask() {
             @Override
-            public Integer call() throws Exception {
-                return waitNotifyService3.getTotalParkedOperationCount();
+            public void run() {
+                assertEquals(1, waitNotifyService3.getTotalParkedOperationCount());
             }
-        }, 1);
+        });
 
-        action.run(node1, node2, node3);
+        action.run(getNode(hz1), getNode(hz2), getNode(hz3));
 
         // create a new node to prevent same partition assignments
         // after node3 rejoins
@@ -154,18 +141,16 @@ public class Invocation_NetworkSplitTest extends HazelcastTestSupport {
 
         assertTrueEventually(new AssertTask() {
             @Override
-            public void run() throws Exception {
-                Assert.assertEquals(0, node1.partitionService.getMigrationQueueSize());
+            public void run() {
+                assertEquals(0, getPartitionService(hz1).getMigrationQueueSize());
             }
         });
 
         // Let node3 detect the split and merge it back to other two.
-        ClusterServiceImpl clusterService3 = node3.getClusterService();
-        clusterService3.merge(node1.address);
+        ClusterServiceImpl clusterService3 = (ClusterServiceImpl) getClusterService(hz3);
+        clusterService3.merge(getAddress(hz1));
 
-        assertEquals(4, node1.getClusterService().getSize());
-        assertEquals(4, node2.getClusterService().getSize());
-        assertEquals(4, node3.getClusterService().getSize());
+        assertClusterSizeEventually(4, hz1, hz2, hz3);
 
         assertEquals(0, waitNotifyService3.getTotalParkedOperationCount());
     }

